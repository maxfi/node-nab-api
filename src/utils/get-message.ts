import { isPlainObject } from 'is-plain-object';

const convert = (x: object) => {
  if (isPlainObject(x)) {
    return Object.entries(x).reduce((acc, value) => {
      const [k, v] = value;
      acc[k] = convert(v);
      return acc;
    }, {} as Record<string, unknown>);
  }

  return { _text: x };
};

export const getMessage = (x: object) => {
  if (!isPlainObject(x)) throw new Error('Requires a plain object');
  return convert(x);
};
