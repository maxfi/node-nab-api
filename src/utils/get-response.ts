import { isPlainObject } from 'is-plain-object';

const convert = (x: any) => {
  if (!isPlainObject(x)) return x;
  if (x._text) return x._text;
  return Object.entries(x).reduce((acc, value) => {
    const [k, v] = value;
    acc[k] = convert(v);
    return acc;
  }, {} as Record<string, string>);
};

export const getResponse = (x: object) => {
  if (!isPlainObject(x)) throw new Error('Requires a plain object');
  return convert(x);
};
