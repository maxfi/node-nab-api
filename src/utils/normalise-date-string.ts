import parse from 'date-fns/parse';

const pad = (str: string) => (str.length === 1 ? '0' + str : str);

export const normaliseDateString = (str: string) => {
  const [d, o] = str.split('000+');
  const offsetMins = parseInt(o, 10);
  const hours = pad(Math.floor(offsetMins / 60).toString());
  const mins = pad((offsetMins % 60).toString());
  const fixed = `${d}+${hours}:${mins}`;
  return parse(fixed, 'yyyyddMMHHmmssSSSXXX', new Date()).toISOString();
};
