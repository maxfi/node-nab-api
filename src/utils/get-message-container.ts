import { getMessage } from './get-message';

export const getMessageContainer = (input: {
  messageId: string;
  credentials: { merchantId: string; password: string };
  requestType: string;
  dataElementName: string;
  payload: object;
  options: { timeout?: string };
}) => {
  input.options = input.options || {};
  return {
    _declaration: {
      _attributes: {
        version: '1.0',
        encoding: 'UTF-8',
      },
    },
    NABTransactMessage: {
      MessageInfo: {
        messageID: {
          _text: input.messageId,
        },
        messageTimestamp: {
          _text: new Date().toISOString(),
        },
        timeoutValue: {
          _text: input.options.timeout || '60',
        },
        apiVersion: {
          _text: 'spxml-3.0',
        },
      },
      MerchantInfo: {
        merchantID: {
          _text: input.credentials.merchantId,
        },
        password: {
          _text: input.credentials.password,
        },
      },
      RequestType: {
        _text: input.requestType,
      },
      [input.dataElementName]: {
        [input.dataElementName + 'List']: {
          _attributes: {
            count: '1',
          },
          [input.dataElementName + 'Item']: Object.assign(
            { _attributes: { ID: '1' } },
            getMessage(input.payload)
          ),
        },
      },
    },
  };
};
