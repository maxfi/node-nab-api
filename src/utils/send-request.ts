import fetch from 'node-fetch';
import { js2xml, xml2js } from 'xml-js';
import { getResponse } from './get-response';

export const toXml = (x: object) => js2xml(x, { spaces: 4, compact: true });
export const fromXml = (str: string) => xml2js(str, { compact: true });

/**
 * Make the request to the NAB payment gateway API.
 * Converts the payload to XML and converts the response to Object.
 * @param  url - Endpoint URL.
 * @param  payload - Request body.
 * @return Returns the API response as Object.
 */
export const sendRequest = async (url: string, payload: object) => {
  const xml = toXml(payload);
  const response = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'text/xml' },
    body: xml,
  });
  if (!response.ok) {
    const { status, statusText } = response;
    console.log('REQUEST:', url, ':', xml);
    throw new Error(`Bad response from NAB API: ${status} - ${statusText}`);
  }
  const body = await response.text();
  return getResponse(fromXml(body));
};
