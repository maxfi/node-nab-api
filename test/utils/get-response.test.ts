import { getResponse } from '../../src/utils/get-response';
import { fromXml } from '../../src/utils/send-request';
import { readFileFromPath } from '../helpers/read-file-from-path';

test('getResponse', async () => {
  const response = getResponse(
    fromXml(await readFileFromPath('test/fixtures/add-token-response.xml'))
  );
  expect(response).toMatchSnapshot();
});
