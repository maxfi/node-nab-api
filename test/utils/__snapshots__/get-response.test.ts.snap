// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getResponse 1`] = `
Object {
  "NABTransactMessage": Object {
    "MerchantInfo": Object {
      "merchantID": "XYZ00",
    },
    "MessageInfo": Object {
      "apiVersion": "spxml-3.0",
      "messageID": "8af793f9af34bea0ecd7eff711c9d3",
      "messageTimestamp": "20140710115921873000+600",
    },
    "RequestType": "addToken",
    "Status": Object {
      "statusCode": "0",
      "statusDescription": "Normal",
    },
    "Token": Object {
      "TokenList": Object {
        "TokenItem": Object {
          "CreditCardInfo": Object {
            "cardDescription": "Visa",
            "cardType": "6",
            "expiryDate": "11/15",
            "pan": "444433XXXXXXX111",
          },
          "_attributes": Object {
            "ID": "1",
          },
          "amount": "100",
          "responseCode": "00",
          "responseText": "Successful",
          "successful": "yes",
          "tokenValue": "1234123412341234",
          "transactionReference": Object {},
        },
        "_attributes": Object {
          "count": "1",
        },
      },
    },
  },
  "_declaration": Object {
    "_attributes": Object {
      "encoding": "UTF-8",
      "version": "1.0",
    },
  },
}
`;
